File: 백준/Silver/11724. 연결 요소의 개수/연결 요소의 개수.py
import sys

sys.setrecursionlimit(10**9)

input = sys.stdin.readline


def main(dic, n):
    visited = [False] * (n + 1)

    def dfs(start):
        for nn in dic[start]:
            if not visited[nn]:
                visited[nn] = True
                dfs(nn)

    cnt = 0
    for i in dic.keys():
        if not visited[i]:
            cnt += 1
            visited[i] = True
            dfs(i)

    return cnt


n, m = map(int, input().split())

dic = {i + 1: [] for i in range(n)}
for _ in range(m):
    a, b = map(int, input().split())
    dic[a].append(b)
    dic[b].append(a)

print(main(dic, n))

File: 백준/Silver/11724. 연결 요소의 개수/연결 요소의 개수.cc
#include <iostream>
#include <queue>
#include <vector>
#include <algorithm>

using namespace std;

int n, m;
vector<vector<int>> dic;
vector<int> visited;

void dfs(int start);

int main()
{
  cin >> n >> m;
  dic.resize(n + 1);
  for (size_t i = 0; i < m; i++)
  {
    int a, b;
    cin >> a >> b;
    dic[a].push_back(b);
    dic[b].push_back(a);
  }
  visited.assign(n + 1, 0);

  int cnt = 0;
  for (size_t i = 1; i <= n; i++)
  {
    if (!visited[i])
    {
      cnt += 1;
      visited[i] = 1;
      dfs(i);
    }
  }

  cout << cnt << endl;
}

void dfs(int start)
{
  for (auto &&nv : dic[start])
  {
    if (!visited[nv])
    {
      visited[nv] = 1;
      dfs(nv);
    }
  }
}
